core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/TupleExprs.java

```diff
@@ -60,6 +62,33 @@ public static boolean containsSubquery(TupleExpr t) {
 		return false;
 	}
 
+	/**
+	 * Verifies if the supplied {@link TupleExpr} contains a {@link Extension}. If the supplied TupleExpr is a
+	 * {@link Join} or contains a {@link Join}, a {@link Service} clause or a subquery element, extensions inside that
+	 * element's argument will not be taken into account.
+	 *
+	 * @param t a tuple expression.
+	 * @return <code>true</code> if the TupleExpr contains an Extension (outside of a Join, Service clause, or
+	 *         subquery), <code>false</code> otherwise.
+	 */
+	public static boolean containsExtension(TupleExpr t) {
+		Deque<TupleExpr> queue = new ArrayDeque<>();
+		queue.add(t);
+		while (!queue.isEmpty()) {
+			TupleExpr n = queue.removeFirst();
+			if (n instanceof Extension) {
+				return true;
+			} else if (n instanceof Join
+					|| (n instanceof Projection && ((Projection) n).isSubquery())
+					|| (n instanceof Service)) {
+				return false;
+			} else {
+				queue.addAll(getChildren(n));
+			}
+		}
+		return false;
+	}
+
```

Comment:
Creating an ArrayDeque is fairly expensive. I've optimized a similar method (on the develop branch). Could you rewrite this method to match that approach?

```
public static boolean containsSubquery(TupleExpr t) {

		TupleExpr n = t;
		Deque<TupleExpr> queue = null;

		do {
			if (n instanceof Projection && ((Projection) n).isSubquery()) {
				return true;
			} else if (n instanceof Join) {
				// projections already inside a Join need not be
				// taken into account
				return false;
			} else {
				List<TupleExpr> children = getChildren(n);
				if (!children.isEmpty()) {
					if (queue == null) {
						queue = new ArrayDeque<>();
					}
					queue.addAll(children);
				}
			}

			n = queue != null ? queue.poll() : null;

		} while (n != null);

		return false;
	}
```